### DISCUSSION TITLE
Passport Strategy: ResponseBodyError

### DISCUSSION THREAD

**Discussion by robireton** [panva/openid-client#725] [Q&A] [answered]
I’m trying to migrate from version 5 to 6 and use the included Passport strategy.

Starting with the provided passport example code, I get a `ResponseBodyError` in `Strategy.authorizationCodeGrant` when it calls `client.authorizationCodeGrant`.

I notice that one of the parameters to `client.authorizationCodeGrant` is produced by `Strategy.authorizationCodeGrantParameters`, but that method can only return an empty object.

Can the included Passport strategy work if this is the case?

**Comment by panva** [comment]
What's the [cause] of your ResponseBodyError?

**Reply by robireton** [reply]
Full output is
```
2024-11-20T19:22:25.156Z	10.97.29.245	GET	/
2024-11-20T19:22:25.176Z	10.97.29.245	GET	/login
express deprecated req.host: Use req.hostname instead file:/Users/iretonrt/Projects/Sample/node_modules/openid-client/build/passport.js:151:49
2024-11-20T19:22:32.778Z	10.97.29.245	GET	/oauth2/redirect/google	{"code":"4/0AeanS0YD3uB_DlHxdRrka1ZmfrRieRGDT9PGckavsljqGez_P2yQUm_WRb6EGLoKWTv0PQ","scope":"email openid https://www.googleapis.com/auth/userinfo.email","authuser":"0","prompt":"consent"}
ResponseBodyError: server responded with an error in the response body
    at processGenericAccessTokenResponse (file:///Users/iretonrt/Projects/Sample/node_modules/oauth4webapi/build/index.js:1121:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async processAuthorizationCodeOAuth2Response (file:///Users/iretonrt/Projects/Sample/node_modules/oauth4webapi/build/index.js:1363:20)
    at async Module.authorizationCodeGrant (file:///Users/iretonrt/Projects/Sample/node_modules/openid-client/build/index.js:670:18)
    at async Strategy.authorizationCodeGrant (file:///Users/iretonrt/Projects/Sample/node_modules/openid-client/build/passport.js:121:28)
```

The source that generated it is attached. I’m authenticating with Google.

* [index.js.gz](https://github.com/user-attachments/files/17835799/index.js.gz)

**Reply by panva** [reply]
Thanks, but that doesn't answer my question, please handle the error and log its `cause`.

**Reply by panva** [reply]
```
express deprecated req.host: Use req.hostname instead file:/Users/iretonrt/Projects/Sample/node_modules/openid-client/build/passport.js:151:49
```

this also tells me you probably just need to either

1) upgrade to express@5;r or
2) overload the strategy's `currentUrl` method to return the correct current URL

Your problem is very likely the same as in #714

**Reply by robireton** [reply]
Thank you for your attention and assistance. I didn’t know that Express 5 had been released. I’ll probably just stick with openid-client@5 until I can consider migrating to Express 5. I am still puzzled about how the Passport strategy included with openid-client@6 can work when `Strategy.authorizationCodeGrantParameters` and `Strategy.authorizationRequestParams` both ignore the `options` passed to them and return just `{}`. I would also love to see a working Passport example. The one in the repo doesn’t actually initialize or start Express or use the callbackURL required by Google.

To close the loop on the `cause`: an Error is not being thrown such that my code can catch and examine it. The error is being written to the console and also appears in the browser.

**Reply by panva** [reply, marked as answer]
All you have to do is make `currentUrl` return the current URL of the request incl. its search parameters.

Don't worry about `authorizationCodeGrantParameters` and `authorizationRequestParams`, these are used to provide additional / extra parameter to the respective protocol messages.

To explain, the `currentUrl` prototype method on the Strategy is used to determine the current URL the strategy is executed on, ergo the URL that is used as the `redirect_uri` token endpoint parameter. And in tls-offloading proxy setups it just depends on [express configuration](https://expressjs.com/en/guide/behind-proxies.html) to be in order.

https://github.com/panva/openid-client/blob/7405570e440ff39834d63d6a027a29ecd7dc537d/src/passport.ts#L378-L389

**Reply by robireton** [reply]
For what it’s worth, I am specifying `callbackURL` in my Strategy constructor, so the `currentUrl` shouldn’t be relevant? The error I’m seeing happens in the route handler for that callback URL.

**Reply by panva** [reply]
`callbackURL` from the options is only used as the `redirect_uri` parameter in the authorization request. Using it as a fixed value during the callback phase would defeat its purpose.

**Reply by robireton** [reply]
Thanks again for the extra help and the extra attention. Updating my Strategy instance's `currentUrl` to account for the extra path from my reverse proxy did indeed work. (*Now I wonder how my code using openid-client@5 worked with just a fixed `redirect_uri`?*)

**Reply by panva** [reply]
Glad I could help.
